### DISCUSSION TITLE
Seamless JWKs rotation

### DISCUSSION THREAD

**Discussion by vavsab** [zitadel/zitadel#5690] [Ideas]
Please add an ability to introduce a new Signing Key without using it for signing.
It would allow my microservices to update the JWKs cache without need to query Zitatel on the missing KID. 

Querying on the missing KID complicates microservice logic for updating JWKs and introduces performance issues when we have new KID going through all microservices and triggering JWKs cache upgrade.

Both Auth0 & Keycloak allow this.

Auth0 always presents next key in advance
![image](https://user-images.githubusercontent.com/10233442/232065025-77c61fb5-33a0-455a-bb2d-becf85b14731.png)

Keycloak allows creating a new key, make it visible, but start signing with it when ready
![image](https://user-images.githubusercontent.com/10233442/232065393-38d06c04-9817-4f17-8c27-5c6bd057f0b2.png)

Approach with waiting for new key to propagate [is non-standard](https://github.com/zitadel/zitadel/issues/2041#issuecomment-882458859) but it simplifies clients logic and makes overall system more predictable without any load spikes on JWKs rotation. 

P.S. One more drawback of querying on the missing KID is that a hacker can use JWTs with invalid KID to flood Zitadel with requests.

**Comment by ghstahl** [comment]
Future publishing is to allow things to run smoothly day to day.  It doesn't introduce a security issue for the OAuth2 service.  

There were other discussions around this regarding what happens when a private key gets compromised.  That would result in the KID being removed from the published JWKS.  That is a real case but it isn't in conflict with the smooth operation of future publishing.  

Both parties need to be in concert, and [asp.net cores auth middleware](https://github.com/dotnet/aspnetcore/blob/7a0d0de38dbb729b823115d60d2d9ac691d8c292/src/Security/Authentication/JwtBearer/src/JwtBearerHandler.cs#L121) accounts for it.  

With the current design every day is a compromised day and middleware has to react to not having a KID in the client cache.  
On the smooth operating side, I can absorb a few missed fetches pulling the JWKS if I have many days to pull the future one before its used.   

I vote for smooth operations on the zitadel side, and the responsibility of picking a good middleware that covers all the bases is my responsibility for that rare case of a compromised key.

**Comment by fforootd** [comment]
Just for reference, this discussion in our chat relates to this https://discord.com/channels/927474939156643850/1095955147647103006

**Comment by ghstahl** [comment]
Btw: this is what I use for my key rotation. 

https://github.com/fluffy-bunny/crypto-gen

I cook about 20 years worth and have it injected. Then my jwks only displays the ones that are in a range where an outstanding jwt may need them. If I am compromised I rerun another set and inject them. No code to keep rotating and storing. The whole thing is static.

**Comment by muhlemmer** [comment]
## Proposal

I had some talks with colleagues @livio-a s and @fforootd. Based on those talks I would like to make this proposal:

**Use the API to manage keys** and get rid of time-based magic.

- On instance creation we will create the first keypair
- Have a keys API(v2) resource where the `IAM_OWNER` can:
  - Create a new keypair
  - Set a private key as the current signing key
  - Delete keypairs for forced rotation
- After the API is available we can add key management to the console UI.

If time-based rotation is still required, this can be achieved by a cron jon calling the API. For example zitadel cloud could use [google cloud scheduler](https://cloud.google.com/scheduler). Other deployment targets may offer similar features.

I'll just leave this here for some feedback and will create feature issues after the discussion is concluded.

**Reply by fcenedes** [reply]
We would welcome such feature :-)

**Reply by fforootd** [reply]
perfect that is nice to hear

**Reply by muhlemmer** [reply]
As this thread was closed, I opened a new one with additional background here:

- https://github.com/zitadel/zitadel/discussions/7464

**Comment by vavsab** [comment]
My company decided to sunset Zitadel. So we do not need this feature anymore.

**Reply by fforootd** [reply]
Thanks for sharing, even though it is sad to hear.

It would be nice if you could share what the ultimate reason was to sunset zitadel. 
Would you mind sharing this with me on florian@zitadel.com?

We try to learn from such cases.

**Reply by vavsab** [reply]
Sure. I have sent you my reasons to your email
